README.md
## Apica Worldmap Panel Plugin for Grafana

Apica Worldmap Panel Plugin supports geo coordinates data provided by Apica Synthetic Monitoring datasource.

Based on [Grafana Worldmap Panel plugin](https://github.com/grafana/worldmap-panel) v0.0.18.

The original readme file can be found here: [README.md](https://github.com/grafana/worldmap-panel/blob/master/README.md).

### Changelog

#### v0.0.1

Based on [Grafana Worldmap Panel plugin](https://github.com/grafana/worldmap-panel) v0.0.18.

The only format from datasource is supported: table.

Required the following 'columns' in table:

  * geohash - encoded latitude and longitude;
  * metric - takes a color code: 0 for gray, 1 for green, 2 for yellow, 3 for orange and 4 for red;
  * a column with popup label text that will be shown on mouse over the point on the map, optional, if nothing then n/a will be shown as a popup label.

Editor tab is simplified. The following settings can be found here:

1. Map Visual Options:
	* Center - for centering the map
	* Initial Zoom
	* Min Circle Size - no different sizes are allowed, so this field sets all circles size
	* Sticky Labels - if checked then popup labels will not disappear automatically

2. Map Data Options:
	* Table Label Field - table column that should be used as popup label text. Predefined with 'summaryHtml' value, which is the column name that contains Apica Syntetic Monitoring location checks summary. 

#### v0.0.2

Upgraded from [Grafana Worldmap Panel plugin](https://github.com/grafana/worldmap-panel) v0.0.21

#### v0.1.0

Minor bugfixes. Version is 0.1.0 now.

#### v0.1.1

Bugfixes:
WPMX-4090 - Redundant options sometimes are not hidden on first load
Fixed path to panel view html's (module.html) to be relative, so plugin will work without changes when included into an app plugin.


CHANGELOG.md
# Changelog

## v0.0.2

- Fixes bug where time series with a country code not found in the country data crashes the panel.
- Adds some extra country codes to the country data to be more similar to the MaxMind Country database.

## v0.0.3

- Support for lowercase country codes for non-elasticsearch datasources.

## v0.0.4

- Fixes snapshotting.

## v.0.0.5

- Adds support for json and jsonp endpoints for location data.

## v.0.0.6

- Adds decimal places option for data values in circle popovers.

## v.0.0.7

- Updates tile map urls to https to avoid mixed content warnings on https sites.

## v.0.0.8

- Saves location data in the dashboard json when snapshotting. This means snapshots should work even when using a custom endpoint for returning a location data json file.

## v.0.0.9

- Fixes bug that meant location data did not refresh after being changed in the editor. It required the page to be refreshed to reload it.

## v.0.0.10

- Performance fix for snapshotting. Sets maxdatapoints to 1 to minimize data that needs to be saved in the snapshot.

## v.0.0.11

- Zoom issue fix and adds a states options for USA states location data.

## v.0.0.12

- Fixes [issue with the JSON endpoint not working](https://github.com/grafana/worldmap-panel/issues/22)

## v.0.0.13

- New location data option -> table data. Location data can now come from data sources other than graphite and Elasticsearch (InfluxDb for example). See table data instructions above on how to use it.

## v.0.0.14

- Various [bug](https://github.com/grafana/worldmap-panel/pull/31) [fixes](https://github.com/grafana/worldmap-panel/pull/32) provided by [linkslice](https://github.com/linkslice) (Thank you!)

## v.0.0.15

- Fix for change in Grafana that [breaks Worldmap panels using Geohash or Table Data](https://github.com/grafana/worldmap-panel/issues/45).

## v.0.0.16

- Option for sticky labels. Fix for https://github.com/grafana/worldmap-panel/issues/27

- Ability to hide null or 0 values. Fix for https://github.com/grafana/worldmap-panel/issues/13

- Background color change. Fixes https://github.com/grafana/worldmap-panel/issues/36

- Dynamic thresholds implemented by [Sam Hatchett](https://github.com/samhatchett). Can now have more than 2 threshold values. Thanks! Fixes https://github.com/grafana/worldmap-panel/issues/25

- Validation and default values for option fields. Fixes https://github.com/grafana/worldmap-panel/issues/29

## v.0.0.17

- Adds Country data with 3-letter country codes.

## v0.0.18

- Fixes some coordinates in the country location data.
- Last Geohash as center for the map - it centers the map on the last geohash position received. Useful for real time tracking (with auto refresh on in Grafana).

## v0.0.19

Fix for Elasticsearch geohash maps after breaking change in Grafana 4.5.

## v0.0.20

Small fixes.

## v0.0.21

 - Support for new data source integration, the Dynamic JSON endpoint [#103](https://github.com/grafana/worldmap-panel/issues/103), thanks [@LostInBrittany](https://github.com/LostInBrittany)
 - Fix for using floats in thresholds [#79](https://github.com/grafana/worldmap-panel/issues/79), thanks [@fabienpomerol](https://github.com/fabienpomerol)
 - Fix for newly created Worldmap panels overflowing their boundaries.
 - Fix for legend css
 - Turned off mouse wheel zoom

## v0.1.2

 - Added Leaflet.markercluster plugin for grouping feature.

src/data_formatter.js
import _ from 'lodash';
import decodeGeoHash from './geohash';

export default class DataFormatter {
  constructor(ctrl, kbn) {
    this.ctrl = ctrl;
    this.kbn = kbn;
  }

  setValues(data) {
    if (this.ctrl.series && this.ctrl.series.length > 0) {
      let highestValue = 0;
      let lowestValue = Number.MAX_VALUE;

      this.ctrl.series.forEach((serie) => {
        const lastPoint = _.last(serie.datapoints);
        const lastValue = _.isArray(lastPoint) ? lastPoint[0] : null;
        const location = _.find(this.ctrl.locations, (loc) => { return loc.key.toUpperCase() === serie.alias.toUpperCase(); });

        if (!location) return;

        if (_.isString(lastValue)) {
          data.push({key: serie.alias, value: 0, valueFormatted: lastValue, valueRounded: 0});
        } else {
          const dataValue = {
            key: serie.alias,
            locationName: location.name,
            locationLatitude: location.latitude,
            locationLongitude: location.longitude,
            value: serie.stats[this.ctrl.panel.valueName],
            valueFormatted: lastValue,
            valueRounded: 0
          };

          if (dataValue.value > highestValue) highestValue = dataValue.value;
          if (dataValue.value < lowestValue) lowestValue = dataValue.value;

          dataValue.valueRounded = this.kbn.roundValue(dataValue.value, parseInt(this.ctrl.panel.decimals, 10) || 0);
          data.push(dataValue);
        }
      });

      data.highestValue = highestValue;
      data.lowestValue = lowestValue;
      data.valueRange = highestValue - lowestValue;
    }
  }

  createDataValue(encodedGeohash, decodedGeohash, locationName, value) {
    const dataValue = {
      key: encodedGeohash,
      locationName: locationName,
      locationLatitude: decodedGeohash.latitude,
      locationLongitude: decodedGeohash.longitude,
      value: value,
      valueFormatted: value,
      valueRounded: 0
    };

    dataValue.valueRounded = this.kbn.roundValue(dataValue.value, this.ctrl.panel.decimals || 0);
    return dataValue;
  }

  setGeohashValues(dataList, data) {
    if (!this.ctrl.panel.esGeoPoint || !this.ctrl.panel.esMetric) return;

    if (dataList && dataList.length > 0) {
      let highestValue = 0;
      let lowestValue = Number.MAX_VALUE;

      dataList.forEach((result) => {
        if (result.type === 'table') {
          const columnNames = {};

          result.columns.forEach((column, columnIndex) => {
            columnNames[column.text] = columnIndex;
          });

          result.rows.forEach((row) => {
            const encodedGeohash = row[columnNames[this.ctrl.panel.esGeoPoint]];
            const decodedGeohash = decodeGeoHash(encodedGeohash);
            const locationName = this.ctrl.panel.esLocationName ? row[columnNames[this.ctrl.panel.esLocationName]] : encodedGeohash;
            const value = row[columnNames[this.ctrl.panel.esMetric]];

            const dataValue = this.createDataValue(encodedGeohash, decodedGeohash, locationName, value, highestValue, lowestValue);
            if (dataValue.value > highestValue) highestValue = dataValue.value;
            if (dataValue.value < lowestValue) lowestValue = dataValue.value;
            data.push(dataValue);
          });

          data.highestValue = highestValue;
          data.lowestValue = lowestValue;
          data.valueRange = highestValue - lowestValue;
        } else {
          result.datapoints.forEach((datapoint) => {
            const encodedGeohash = datapoint[this.ctrl.panel.esGeoPoint];
            const decodedGeohash = decodeGeoHash(encodedGeohash);
            const locationName = this.ctrl.panel.esLocationName ? datapoint[this.ctrl.panel.esLocationName] : encodedGeohash;
            const value = datapoint[this.ctrl.panel.esMetric];

            const dataValue = this.createDataValue(encodedGeohash, decodedGeohash, locationName, value, highestValue, lowestValue);
            if (dataValue.value > highestValue) highestValue = dataValue.value;
            if (dataValue.value < lowestValue) lowestValue = dataValue.value;
            data.push(dataValue);
          });

          data.highestValue = highestValue;
          data.lowestValue = lowestValue;
          data.valueRange = highestValue - lowestValue;
        }
      });
    }
  }

  static tableHandler(tableData) {
    const datapoints = [];

    if (tableData.type === 'table') {
      const columnNames = {};

      tableData.columns.forEach((column, columnIndex) => {
        columnNames[columnIndex] = column.text;
      });

      tableData.rows.forEach((row) => {
        const datapoint = {};

        row.forEach((value, columnIndex) => {
          const key = columnNames[columnIndex];
          datapoint[key] = value;
        });

        datapoints.push(datapoint);
      });
    }

    return datapoints;
  }

  setTableValues(tableData, data) {
    if (tableData && tableData.length > 0) {
      let highestValue = 0;
      let lowestValue = Number.MAX_VALUE;

      tableData[0].forEach((datapoint) => {
        if (!datapoint.geohash) {
          return;
        }

        const encodedGeohash = datapoint.geohash;
        const decodedGeohash = decodeGeoHash(encodedGeohash);

        const dataValue = {
          key: encodedGeohash,
          locationName: datapoint.name,
          summary: datapoint[this.ctrl.panel.tableLabel] || 'n/a',
          countBySeverity: datapoint.countBySeverity,
          locationLatitude: decodedGeohash.latitude,
          locationLongitude: decodedGeohash.longitude,
          value: datapoint.metric,
          valueFormatted: datapoint.metric,
          valueRounded: 0
        };

        if (dataValue.value > highestValue) highestValue = dataValue.value;
        if (dataValue.value < lowestValue) lowestValue = dataValue.value;

        dataValue.valueRounded = this.kbn.roundValue(dataValue.value, this.ctrl.panel.decimals || 0);
        data.push(dataValue);
      });

      data.highestValue = highestValue;
      data.lowestValue = lowestValue;
      data.valueRange = highestValue - lowestValue;
    }
  }

  setJsonValues(data) {
    if (this.ctrl.series && this.ctrl.series.length > 0) {
      let highestValue = 0;
      let lowestValue = Number.MAX_VALUE;

      this.ctrl.series.forEach((point) => {
        const dataValue = {
          key: point.key,
          locationName: point.name,
          locationLatitude: point.latitude,
          locationLongitude: point.longitude,
          value: (point.value !== undefined) ? point.value : 1,
          valueRounded: 0
        };
        if (dataValue.value > highestValue) highestValue = dataValue.value;
        if (dataValue.value < lowestValue) lowestValue = dataValue.value;
        dataValue.valueRounded = Math.round(dataValue.value);
        data.push(dataValue);
      });
      data.highestValue = highestValue;
      data.lowestValue = lowestValue;
      data.valueRange = highestValue - lowestValue;
    }
  }
}



src/worldmap_apica.js
import WorldMapOriginal from './worldmap';

const tileServers = {
  'CartoDB Positron': { url: 'https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png', attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a> &copy; <a href="http://cartodb.com/attributions">CartoDB</a>', subdomains: 'abcd'},
  'CartoDB Dark': {url: 'https://cartodb-basemaps-{s}.global.ssl.fastly.net/dark_all/{z}/{x}/{y}.png', attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a> &copy; <a href="http://cartodb.com/attributions">CartoDB</a>', subdomains: 'abcd'}
};

export default class WorldMap extends WorldMapOriginal {
  // in table Location Data mode locationName is the value of the field which name is set in "Table Label Field".
  // can be set to any field from table datasource, so we just show it instead of the original that shows more than we want.
  createPopup(circle, locationName, value) {
    super.createPopup(circle, locationName, value);
    circle.unbindPopup();
    circle.bindPopup(locationName, {'offset': window.L.point(0, -7), 'className': 'worldmap-popup', 'closeButton': this.ctrl.panel.stickyLabels});
  }

  createMap() {
    window.L.Icon.Default.imagePath = 'public/img';
    this.markers = [];
    this.legend = window.L.control({position: 'bottomleft'});
    const mapCenter = window.L.latLng(parseFloat(this.ctrl.panel.mapCenterLatitude), parseFloat(this.ctrl.panel.mapCenterLongitude));
    this.map = window.L.map(this.mapContainer, {worldCopyJump: true, center: mapCenter})
      .fitWorld()
      .zoomIn(parseInt(this.ctrl.panel.initialZoom, 10));
    this.map.panTo(mapCenter);
    this.map.scrollWheelZoom.disable();

    const selectedTileServer = tileServers[this.ctrl.tileServer];
    window.L.tileLayer(selectedTileServer.url, {
      maxZoom: 18,
      subdomains: selectedTileServer.subdomains,
      reuseTiles: true,
      detectRetina: true,
      attribution: selectedTileServer.attribution
    }).addTo(this.map);
  }

  createCircles(data) {
    var self = this;

    var markers = window.L.markerClusterGroup({
      showCoverageOnHover: false,
      iconCreateFunction: function (cluster) {
        var markers = cluster.getAllChildMarkers();
        var severityClusterClass = self.getClassWithWorstSeverity(markers);

        return L.divIcon({ html: `<div><span>${markers.length}</span></div>`, className: `leaflet-marker-icon marker-cluster ${severityClusterClass} leaflet-zoom-animated leaflet-clickable`, iconSize: L.point(40, 40) });
      },
    });
    
    data.forEach((dataPoint) => {
      var marker = new window.L.CircleMarker(new window.L.LatLng(dataPoint.locationLatitude, dataPoint.locationLongitude), {
        radius: this.calcCircleSize(dataPoint.value || 0),
        color: this.getColor(dataPoint.value),
        fillColor: this.getColor(dataPoint.value),
        fillOpacity: 0.5,
        locationName: dataPoint.locationName,
        countBySeverity: dataPoint.countBySeverity
      });

      this.createPopup(marker, dataPoint.summary, dataPoint.valueRounded);

      this.markers.push(markers.addLayer(marker));
    });

    


    markers.on('clustermouseover', function(e) {
      var locationPopupPartials = '';
      var total = {i: 0, w: 0, e: 0, f: 0, u: 0, total: 0};
      var cluster = self.getClusterFromLayer(e);
      var children = e.layer.getAllChildMarkers();
      
      for (var i = 0; i < children.length; i++) {
        var child = children[i];

        child.options.countBySeverity.total = 
          child.options.countBySeverity.i + 
          child.options.countBySeverity.w + 
          child.options.countBySeverity.e + 
          child.options.countBySeverity.f + 
          child.options.countBySeverity.u;

        locationPopupPartials += self.createLocationPopupPartial(child.options.locationName, child.options.countBySeverity, i < children.length - 1);

        total.i += child.options.countBySeverity.i;
        total.w += child.options.countBySeverity.w;
        total.e += child.options.countBySeverity.e;
        total.f += child.options.countBySeverity.f;
        total.u += child.options.countBySeverity.u;
        total.total += child.options.countBySeverity.total; 
      };

      var label = '<div class="popup-content">';
      label += self.createSummaryPopupPartial('Summary', total);
      label += locationPopupPartials
      label += '</div>';

      cluster.unbindPopup();
      cluster.bindPopup(label, { 'offset': window.L.point(0, -20), 'className': 'worldmap-popup', 'closeButton': self.ctrl.panel.stickyLabels}).openPopup();
    })

    markers.on('clustermouseout', function(e) {
      if (!self.ctrl.panel.stickyLabels) {
        var cluster = self.getClusterFromLayer(e);
    	
        cluster.closePopup();
      }
    })

    this.markerLayer = this.map.addLayer(markers);
  }

  clearCircles() {
    for (var i =0; i < this.markers.length; i++) {
      this.markers[i].off('mouseover');
      this.markers[i].off('mouseout');
      this.map.removeLayer(this.markers[i]);
    }

    if (this.markerLayer) {
      this.map.removeLayer(this.markerLayer);
    }

    this.markers = [];
  }

  createSummaryPopupPartial(name, total) {
    return `<div class="summary-section">
              <div class="summary-section__name">
                ${name}
              </div>
              <div>
                <span class="summary-section__description summary-section__description_total">Total</span>
                <span class="summary-section__description summary-section__description_information">I</span>
                <span class="summary-section__description summary-section__description_warning">W</span>
                <span class="summary-section__description summary-section__description_error">E</span>
                <span class="summary-section__description summary-section__description_fatal">F</span>
                <span class="summary-section__description summary-section__description_unknow">U</span>
              </div>
              <div>
                <span class="summary-section__severity summary-section__severity_total">${total.total}</span>
                <span class="summary-section__severity summary-section__severity_information">${total.i}</span>
                <span class="summary-section__severity summary-section__severity_warning">${total.w}</span>
                <span class="summary-section__severity summary-section__severity_error">${total.e}</span>
                <span class="summary-section__severity summary-section__severity_fatal">${total.f}</span>
                <span class="summary-section__severity summary-section__severity_unknow">${total.u}</span>
              </div>
              <hr>
            </div>`;
  }

  createLocationPopupPartial(name, total, needHr) {
    var hrBlock = needHr ? '<hr>' : '';

    return `<div class="location-section">
              <div class="location-section__name">
                ${name}
              </div>
              <div>
                <span class="location-section__severity location-section__severity_total">${total.total}</span>
                <span class="location-section__severity location-section__severity_information">${total.i}</span>
                <span class="location-section__severity location-section__severity_warning">${total.w}</span>
                <span class="location-section__severity location-section__severity_error">${total.e}</span>
                <span class="location-section__severity location-section__severity_fatal">${total.f}</span>
                <span class="location-section__severity location-section__severity_unknow">${total.u}</span>
              </div>
              ${hrBlock}
            </div>`;
  }

  getClusterFromLayer(e) {
    var marker = e.layer.getAllChildMarkers()[0];
    
    return e.target.getVisibleParent(marker);
  }

  getClassWithWorstSeverity(markers) {
    var totals = {i: 0, w: 0, e: 0, f: 0, u: 0};
    for (var i = 0; i < markers.length; i++) {
      var countBySeverity = markers[i].options.countBySeverity;

      totals.i += countBySeverity.i;
      totals.w += countBySeverity.w;
      totals.e += countBySeverity.e;
      totals.f += countBySeverity.f;
      totals.u += countBySeverity.u;
    }

    if (totals.f > 0) return 'marker-cluster_fatal';
    if (totals.e > 0) return 'marker-cluster_error';
    if (totals.w > 0) return 'marker-cluster_warning';
    if (totals.i > 0) return 'marker-cluster_information';
    if (totals.u > 0) return 'marker-cluster_unknown';

    return '';
  }
}

src/worldmap_ctrl_apica.js
import WorldmapCtrlOriginal from './worldmap_ctrl';

import $ from 'jquery';

const panelDefaultsApica = {
  locationData: 'table',
  tableLabel: 'summaryHtml',
  circleMinSize: 5,
  circleMaxSize: 5,
  showLegend: false,
  stickyLabels: true,
  thresholds: '1,2,3,4',
  colors: [
    'rgba(90, 98, 90, 0.9)',
    'rgba(50, 172, 45, 0.97)',
    'rgba(224, 210, 0, 0.97)',
    'rgba(237, 129, 40, 0.89)',
    'rgba(245, 54, 54, 0.9)'
  ]
};

const editTabIndex = 2;

export default class WorldmapCtrl extends WorldmapCtrlOriginal {

  constructor($scope, $injector, contextSrv) {
    super($scope, $injector, contextSrv);

    // tableLabel is not initiated in parent by default, so this is the mark that here we should init Apica checks summary by location format. 
    if (!this.panel.tableLabel) {
      _.assign(this.panel, _.cloneDeep(panelDefaultsApica));
    }
  }

  onDataSnapshotLoad(snapshotData) {
    super.onDataSnapshotLoad(snapshotData);
    // a hack to make sure that map is initialized (it doesn't happen in time when work in snapshot mode) 
    if (!this.map) {
      setTimeout(() => {
        this.render();
      }, 100);
    }
  }

  onInitEditMode() {
    // tab will be added as a tag with name 'panel-editor-tab-{pluginId}{editTabIndex}', for example: <panel-editor-tab-apica-worldmap-panel2> 
    this.addEditorTab('Worldmap', `public/plugins/${this.pluginId}/partials/editor.html`, editTabIndex);
  }

  initCtrl() {
    let editTabElementName = `panel-editor-tab-${this.pluginId}${editTabIndex}`;
    // hide not useless Map Visual Options:
    // - Max Circle Size - circle sizes should be the same for now
    // - Decimals - metric is used for the color, so decimals settings is useless
    // - Unit - same as Decimals
    // - Show Legend - legend shows metric values, but they as color codes are useless
    // - Location Data - should always be 'table' (ASM Datasource provides data in table format)
    // - Aggregation - not used, 'metric' property from data source is always used
    $(`.tabbed-view-body ${editTabElementName}`)
      .find(`
        input[ng-model="ctrl.panel.circleMaxSize"],
        input[ng-model="ctrl.panel.decimals"],
        input[ng-model="ctrl.panel.unitSingular"],
        gf-form-switch[checked="ctrl.panel.showLegend"],
        select[ng-model="ctrl.panel.locationData"],
        select[ng-model="ctrl.panel.valueName"]
      `).closest('.gf-form').hide();

    // hide not useless Map Visual Options:
    // - Threshold Options (Thresholds, Colors) - thresholds are used for circles coloring that 
    // - Hide series (With only nulls, With only zeros)
    $(`.tabbed-view-body ${editTabElementName}`)
      .find(`
        input[ng-model="ctrl.panel.thresholds"],
        gf-form-switch[checked="ctrl.panel.hideEmpty"]
      `).closest('.gf-form-group').hide();
    
    $(`.tabbed-view-body ${editTabElementName} input[ng-model="ctrl.panel.circleMinSize"]`).siblings('.gf-form-label').text('Circle Size');
    
    // Editor is hidden by default. Make it visible.
    $(`.tabbed-view-body ${editTabElementName} .editor-row`).show();
  }
  
  render() {
    // different circle sizes in current implementation has no sence, so we adjust them to be equal.
    this.panel.circleMaxSize = this.panel.circleMinSize;
    super.render();
  }
}

// note that plugin id (apica-worldmap-panel) is hardcoded here as templateUrl is static
// it isn't clear how to avoid hardcode (like it was done in the constructor using $scope.ctrl.pluginId)
// from grafana hosting all plugins will be available from linear structure, no matter how they actually added into physical plugins folder.
// this way ('../../plugins/pluginid/resource) we make sure plugin will work even if it's included in an app plugin sub-folder.  
WorldmapCtrl.templateUrl = '../../plugins/apica-worldmap-panel/module.html';

src/worldmap_ctrl.js
/* eslint import/no-extraneous-dependencies: 0 */
import {MetricsPanelCtrl} from 'app/plugins/sdk';
import TimeSeries from 'app/core/time_series2';
import kbn from 'app/core/utils/kbn';

import _ from 'lodash';
import mapRenderer from './map_renderer';
import DataFormatter from './data_formatter';
import './css/worldmap-panel.css!';

const panelDefaults = {
  maxDataPoints: 1,
  mapCenter: '(0°, 0°)',
  mapCenterLatitude: 0,
  mapCenterLongitude: 0,
  initialZoom: 1,
  valueName: 'total',
  circleMinSize: 2,
  circleMaxSize: 30,
  locationData: 'countries',
  thresholds: '0,10',
  colors: ['rgba(245, 54, 54, 0.9)', 'rgba(237, 129, 40, 0.89)', 'rgba(50, 172, 45, 0.97)'],
  unitSingle: '',
  unitPlural: '',
  showLegend: true,
  esMetric: 'Count',
  decimals: 0,
  hideEmpty: false,
  hideZero: false,
  stickyLabels: false
};

const mapCenters = {
  '(0°, 0°)': {mapCenterLatitude: 0, mapCenterLongitude: 0},
  'North America': {mapCenterLatitude: 40, mapCenterLongitude: -100},
  'Europe': {mapCenterLatitude: 46, mapCenterLongitude: 14},
  'West Asia': {mapCenterLatitude: 26, mapCenterLongitude: 53},
  'SE Asia': {mapCenterLatitude: 10, mapCenterLongitude: 106},
  'Last GeoHash': {mapCenterLatitude: 0, mapCenterLongitude: 0}
};

export default class WorldmapCtrl extends MetricsPanelCtrl {
  constructor($scope, $injector, contextSrv) {
    super($scope, $injector);

    this.setMapProvider(contextSrv);
    _.defaults(this.panel, panelDefaults);

    this.dataFormatter = new DataFormatter(this, kbn);

    this.events.on('init-edit-mode', this.onInitEditMode.bind(this));
    this.events.on('data-received', this.onDataReceived.bind(this));
    this.events.on('panel-teardown', this.onPanelTeardown.bind(this));
    this.events.on('data-snapshot-load', this.onDataSnapshotLoad.bind(this));

    this.loadLocationDataFromFile();
  }

  setMapProvider(contextSrv) {
    this.tileServer = contextSrv.user.lightTheme ? 'CartoDB Positron' : 'CartoDB Dark';
    this.setMapSaturationClass();
  }

  setMapSaturationClass() {
    if (this.tileServer === 'CartoDB Dark') {
      this.saturationClass = 'map-darken';
    } else {
      this.saturationClass = '';
    }
  }

  loadLocationDataFromFile(reload) {
    if (this.map && !reload) return;

    if (this.panel.snapshotLocationData) {
      this.locations = this.panel.snapshotLocationData;
      return;
    }

    if (this.panel.locationData === 'jsonp endpoint') {
      if (!this.panel.jsonpUrl || !this.panel.jsonpCallback) return;

      window.$.ajax({
        type: 'GET',
        url: this.panel.jsonpUrl + '?callback=?',
        contentType: 'application/json',
        jsonpCallback: this.panel.jsonpCallback,
        dataType: 'jsonp',
        success: (res) => {
          this.locations = res;
          this.render();
        }
      });
    } else if (this.panel.locationData === 'json endpoint') {
      if (!this.panel.jsonUrl) return;

      window.$.getJSON(this.panel.jsonUrl).then((res) => {
        this.locations = res;
        this.render();
      });
    } else if (this.panel.locationData === 'table') {
      // .. Do nothing
    } else if (this.panel.locationData !== 'geohash' && this.panel.locationData !== 'json result') {
      window.$.getJSON(`public/plugins/${this.pluginId}/data/` + this.panel.locationData + '.json')
        .then(this.reloadLocations.bind(this));
    }
  }

  reloadLocations(res) {
    this.locations = res;
    this.refresh();
  }

  onPanelTeardown() {
    if (this.map) this.map.remove();
  }

  onInitEditMode() {
    this.addEditorTab('Worldmap', `public/plugins/${this.pluginId}/partials/editor.html`, 2);
  }

  onDataReceived(dataList) {
    if (!dataList) return;

    if (this.dashboard.snapshot && this.locations) {
      this.panel.snapshotLocationData = this.locations;
    }

    const data = [];

    if (this.panel.locationData === 'geohash') {
      this.dataFormatter.setGeohashValues(dataList, data);
    } else if (this.panel.locationData === 'table') {
      const tableData = dataList.map(DataFormatter.tableHandler.bind(this));
      this.dataFormatter.setTableValues(tableData, data);
    } else if (this.panel.locationData === 'json result') {
      this.series = dataList;
      this.dataFormatter.setJsonValues(data);
    } else {
      this.series = dataList.map(this.seriesHandler.bind(this));
      this.dataFormatter.setValues(data);
    }
    this.data = data;

    this.updateThresholdData();

    if (this.data.length && this.panel.mapCenter === 'Last GeoHash') {
      this.centerOnLastGeoHash();
    } else {
      this.render();
    }
  }

  centerOnLastGeoHash() {
    mapCenters[this.panel.mapCenter].mapCenterLatitude = _.last(this.data).locationLatitude;
    mapCenters[this.panel.mapCenter].mapCenterLongitude = _.last(this.data).locationLongitude;
    this.setNewMapCenter();
  }

  onDataSnapshotLoad(snapshotData) {
    this.onDataReceived(snapshotData);
  }

  seriesHandler(seriesData) {
    const series = new TimeSeries({
      datapoints: seriesData.datapoints,
      alias: seriesData.target,
    });

    series.flotpairs = series.getFlotPairs(this.panel.nullPointMode);
    return series;
  }

  setNewMapCenter() {
    if (this.panel.mapCenter !== 'custom') {
      this.panel.mapCenterLatitude = mapCenters[this.panel.mapCenter].mapCenterLatitude;
      this.panel.mapCenterLongitude = mapCenters[this.panel.mapCenter].mapCenterLongitude;
    }
    this.mapCenterMoved = true;
    this.render();
  }

  setZoom() {
    this.map.setZoom(this.panel.initialZoom || 1);
  }

  toggleLegend() {
    if (!this.panel.showLegend) {
      this.map.removeLegend();
    }
    this.render();
  }

  toggleStickyLabels() {
    this.map.clearCircles();
    this.render();
  }

  changeThresholds() {
    this.updateThresholdData();
    this.map.legend.update();
    this.render();
  }

  updateThresholdData() {
    this.data.thresholds = this.panel.thresholds.split(',').map((strValue) => {
      return Number(strValue.trim());
    });
    while (_.size(this.panel.colors) > _.size(this.data.thresholds) + 1) {
      // too many colors. remove the last one.
      this.panel.colors.pop();
    }
    while (_.size(this.panel.colors) < _.size(this.data.thresholds) + 1) {
      // not enough colors. add one.
      const newColor = 'rgba(50, 172, 45, 0.97)';
      this.panel.colors.push(newColor);
    }
  }

  changeLocationData() {
    this.loadLocationDataFromFile(true);

    if (this.panel.locationData === 'geohash') {
      this.render();
    }
  }

/* eslint class-methods-use-this: 0 */
  link(scope, elem, attrs, ctrl) {
    mapRenderer(scope, elem, attrs, ctrl);
  }
}

WorldmapCtrl.templateUrl = 'module.html';
